[mono] Update SDKs+roslyn to track `release/3.1.1xx`

radical opened the pull request:
.. missed some bits in the previous commit. Listing the changed versions
again:

```
Microsoft.NET.Sdk.Web:                  3.1.100-preview2.19521.2
Microsoft.NET.Sdk.Razor:                3.1.0-preview2.19518.7

Microsoft.NETCore.App:                  3.1.0-preview2.19521.15
Microsoft.DotNet.Cli.Runtime:           3.1.100-preview2.19521.2

NuGet.Build.Tasks:                      5.3.0-rtm.6251

Roslyn:                                 3.4.0-beta3-19521-01
```

Based on `dotnet/toolset`'s `release/3.1.1xx` branch HEAD
(172f26ddc1dc3c96f14bfcba8519544d30d6d3e2).

- And merge upstream `vs16.4` branch

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/arcade build 20190924.3

- Microsoft.DotNet.Arcade.Sdk - 1.0.0-beta.19474.3

rainersigwald committed:
Update NuGet.config

rainersigwald committed:
Point to 4.8 SDK in amd64 app.config as well (#4787)

This mirrors #4487 to the amd64 app.config.

rainersigwald committed:
Allow Before/AfterTargets for dynamic sln targets (#4694)

Fixes #4692 by ensuring that the targets that will be created exist at evaluation time, too.

treymerkley committed:
Demote corrupt resgen output warning (#4776)

Remove warning that a file would be deleted in the case where it was successfully deleted and would be recreated.

Fixes #4648.

ccastanedaucf committed:
[Static Graph] Fix hangs and redundant code in ParallelWorkSet (#4508)

* Add unit tests

* Fix work set not completing after exception

* Simplify completion logic

* Throw when attempting to add work post-completion

* Ensure fast cancel on cancellation request

* Simplify work set processing API

* Make completion setter private

* Clarify test names

Forgind committed:
Unhandled exception when pipe to worker node closes suddenly (#4790)

If a worker experienced a pipe close between when it recieved a handshake and when it sent a response, it crashed MSBuild. This turns an unhandled exception into a clean error message. Fixes #4768

rainersigwald committed:
Revert "Merge pull request #4733 from jcagme/azdoPublish"

This reverts commit ec061db1c7ea241dd745ff164f1213ba19a3cb7c, reversing
changes made to 8bf929b2096df93a5672a491e4d6277ed3d61e66.

rainersigwald committed:
Merge commit 'db0c839' into darc-vs16.3-4d187365-e5bb-4a76-b00c-10c4b7801617

rainersigwald committed:
Revert SDK update (for servicing 16.3)

rainersigwald committed:
Merge pull request #4796 from microsoft/darc-vs16.3-4d187365-e5bb-4a76-b00c-10c4b7801617

jeffkl committed:
Fix CachingSdkResolverService concurrency. (#4797)

The documentation for ConcurrentDictionary<T>.GetOrAdd() states:

> If you call GetOrAdd simultaneously on different threads, valueFactory may be called multiple times, but only one key/value pair will be added to the dictionary.

In all of my testing, a particular SDK was never resolved more than once so I didn't think we had a problem.  However, on an NGen'd MSBuild, I noticed that an SDK was being resolved multiple times.  This is because the `GetOrAdd()` method does not guarantee that a method is only called once.  Instead, it guarantees that if multiple threads call the method simultaneously they will all receive the same return value.

So the fix is to return a `Lazy<T>` instead.  This means that multiple threads could new up multiple instances of Lazy<T> but all threads will receive the same Lazy<T>.  They will all wait for the same result when accessing `Value` so the SDK will only be resolved once and the value is then cached for subsequent calls.

Fixes #4798

dotnet-maestro[bot] committed:
[master] Update dependencies from dotnet/arcade (#4783)

* Update dependencies from https://github.com/dotnet/arcade build 20191002.6

- Microsoft.DotNet.Arcade.Sdk - 1.0.0-beta.19502.6

* Migrate to Azure Artifacts feed for tools

* Match repo-local version references to 3.0.100

* NuGet packages that get deployed into bootstrap
* the SDK itself

* Include RuntimeIdentifierGraph.json in bootstrap

Avoids errors when building with the bootstrapped build like

```
src/MSBuild/MSBuild.csproj(0,0): error NU1007: The runtime.json specified in the project '/Users/vsts/agent/2.158.0/work/1/s/stage1/bin/bootstrap/netcoreapp2.1/MSBuild/RuntimeIdentifierGraph.json' could not be found.
```

* NuGet 5.3 warnings

* Explicitly specify 4.7.2 in CoreXT package
* Use in-package icon rather than URL

* Use hosted pool

For reasons I don't fully understand, specifying the 'Hosted' pool name
caused the use of different machines, even though it seems like it's
redundantly specifying the default. Removing the specification entirely
assigns to the machines I expect.

rainersigwald committed:
Hardcode channel-publish repo name (#4799)

Work around dotnet/arcade#4087 by hardcoding the repo name in the one channel relevant to 16.3 servicing.

rainersigwald committed:
Switch to VS2019 queue for official build (#4800)

Should fix this error in our official build:

```
##[error].packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19502.6\tools\Tools.proj(0,0): error : Version 3.0.100 of the .NET Core SDK requires at least version 16.3.0 of MSBuild. The current available version of MSBuild is 16.0.461.62831. Change the .NET Core SDK specified in global.json to an older version that requires the MSBuild version currently available.
```

Forgind committed:
Update vs version

Props file initiates the wrong version.

Forgind committed:
Merge pull request #4813 from microsoft/vs-versioning

Deliver v16 props file in VSIX

treymerkley committed:
Remove trailing whitespace in GenerateResource.cs (#4791)

dasMulli committed:
Minor binlog doc change (#4793)

Change `/bl` to `-bl` to follow other places where `/` was replaced with `-` in help texts and documentation.

benvillalobos committed:
Merge vs16.3 into master

jeffkl committed:
Implementation of Directory.Solution.props and Directory.Solution.targets (#4785)

Implementation of https://github.com/microsoft/msbuild/issues/4784 

For the generated metaproj project this imports a file named `Directory.Solution.props` at the top and a file named `Directory.Solution.targets` at the bottom.

This allows users to create a single file that lets them extend the building of all solution files in their repository.

benvillalobos committed:
Merge remote-tracking branch 'upstream/master'

jeffkl committed:
Move Microsoft.Build.Experimental.Graph to Microsoft.Build.Graph (#4812)

Static graph is ready for production use and big partner teams are taking a dependency on it.

benvillalobos committed:
Changing source file extension to protected in CreateManifestResourceName (#4821)

* Changing source file extension to protected

* Built for generated files

Commit:
Delete nuspec files under src/nuget. They are not used anywhere anymore. (#4820)

rainersigwald committed:
Allow case-varying solution targets (#4822)

Fixes #4811 by switching to the concrete implementation type for the _defaultTargetNames set.

As ISet<T>, the overload resolved for Union was from Linq and did not respect the left-hand
set's comparer. By moving the field to the concrete ImmutableHashSet<T> type, the overload
from that type is used, which does use the collection's comparer (case-insensitive).

When the target names are correctly unioned, the null-reference exception is avoided, because
the dummy target is removed only once, instead of twice (throwing NRE the second time).

radical committed:
[mono] Update SDKs+roslyn to track `release/3.1.1xx`

.. missed some bits in the previous commit. Listing the changed versions
again:

```
Microsoft.NET.Sdk.Web:                  3.1.100-preview2.19521.2
Microsoft.NET.Sdk.Razor:                3.1.0-preview2.19518.7

Microsoft.NETCore.App:                  3.1.0-preview2.19521.15
Microsoft.DotNet.Cli.Runtime:           3.1.100-preview2.19521.2

NuGet.Build.Tasks:                      5.3.0-rtm.6251

Roslyn:                                 3.4.0-beta3-19521-01
```

Based on `dotnet/toolset`'s `release/3.1.1xx` branch HEAD
(172f26ddc1dc3c96f14bfcba8519544d30d6d3e2).

radical committed:
Merge remote-tracking branch 'origin/vs16.4' into mono-2019-08

radical commented:
I'm having trouble building this locally due to local network issues, so don't merge this before the build passes.

lewing commented:
DotNetCliVersion looks wrong and it is breaking the logic added to RuntimeIdentifierGraph.json

radical committed:
[mono] Fix build

Recent upstream merge started copying RuntimeIdentifierGraph.json from
dotnet sdk installed as part of the build. Since, for a mono build we
don't install that, this Copy fails. Instead, we keep a copy of
this and related files in the repo and copy them from there.

Also, update the files from sdk 3.0.100 .

radical commented:
Merged after fixing the build.
